### ISSUE TITLE
Editor's selection points to non-existing nodes when value is changed programmatically (Chrome, Safari)

### ISSUE THREAD

**Issue by kamilmielnik** [ianstormtaylor/slate#3878]
#### Do you want to request a _feature_ or report a _bug_?

Bug.

#### What's the current behavior?

![Peek 2020-09-16 19-15](https://user-images.githubusercontent.com/6830683/93335750-1bc29d80-f851-11ea-8e68-6043b3946ef1.gif)

https://codesandbox.io/s/slate-reproductions-forked-lzc7g?file=/index.js

1. Focus the last paragraph in the example
2. Click "Reset" button

`useSlate().selection` points to a non-existing location (location after performing step (1))

Slate: 0.58.4
Browser: Chrome, Safari
OS: Windows, Linux, Mac

#### What's the expected behavior?

`useSlate().selection` points to an existing location. 

Works fine in **Firefox**. 

![Peek 2020-09-16 19-18](https://user-images.githubusercontent.com/6830683/93335997-60e6cf80-f851-11ea-860e-9c8a449f4f10.gif)

**Comment by e1himself** [comment]
@kamilmielnik this issue is no longer reproducible with `slate` dependencies at `~0.66`.

https://codesandbox.io/s/slate-reproductions-forked-ruc80?file=/index.js

![Peek 2021-11-02 13-22](https://user-images.githubusercontent.com/370680/139837713-42f25d5b-d9aa-405d-b6e5-90cef6088650.gif)

So I believe you can close it :tada: :wink:

**Comment by DaniGuardiola** [comment]
Stumbled upon this as well. Crashing production hard. Did you find a usable workaround @kamilmielnik? Got any ideas @e1himself?

I've thought about just resetting the selection, but it is hard to achieve without adding some sort of custom imperative method, since there's no easy way to distinguish between a normal state update and an intentional one.

Thanks!

**Comment by DaniGuardiola** [comment]
We got around it by briefly (`setTimeout(showAgain, 0)`) rendering a placeholder with the same height and style when changing the value programmatically outside of controlled state, in order to force it to remount and reset the selection. It's not great though, since we still get a flash.

**Comment by kamilmielnik** [comment]
> Apologies, the issue still remains in Chrome.

@e1himself 
For the record:
- the issue is reported for **Chrome** & **Safari** (see issue title)
- there is no issue in Firefox since slate@0.58.4 already (see issue description)

> Did you find a usable workaround @kamilmielnik?

@DaniGuardiola 
We've been using this as a workaround: [isSelectionValid](https://github.com/prezly/slate/blob/e89e306419957dad9e787dd0450cd5c3281db8f1/packages/slate-commons/src/commands/isSelectionValid.ts) (you can search that repo for usages of this function).
It does not really address the issue, it just prevents the editor from crashing in those cases.

**Comment by e1himself** [comment]
> I've thought about just resetting the selection, but it is hard to achieve without adding some sort of custom imperative method, since there's no easy way to distinguish between a normal state update and an intentional one.

What if there's an `onChange` handler that will always fix selection if it does not point to an existing node? Do we need to distinguish normal state update and intentional programmatic update in that case?

**Comment by DaniGuardiola** [comment]
@e1himself yep, I think that's the problem with that fix, I don't think there a good way to tell which is which.

@kamilmielnik in our case we have quite a few places where we're accessing the selection, so checking everywhere would be a little messy :/ thx for sharing though!

**Comment by DaniGuardiola** [comment]
@e1himself on a second thought, it might not be necessary to distinguish those, but I'm still worried that it'll be already too late and the editor would've crashed by then, since the value would've been applied already.
